File: EvDemandModel/EvDemandModel/Utils/EVDemandInput.py
from enum import Enum
import json
import shapely

class EVDemandInput:
    def __init__(self,dict=None)->None:
        self.predictorParams: EVDemandInput.PredictorParams
        self.regionData: list[EVDemandInput.RegionData]
        if dict:
            self.__dict__ = dict

    class Polygon:
        def __init__(self,dict=None)->None:
            self.points: list[tuple]
            if dict:
                self.__dict__ = dict

    class RegionType(Enum):
        Dist=0
        Primary=1
        GSP=2

    class RegionData:
        def __init__(self,dict=None)->None:
            self.id: int
            self.type: EVDemandInput.RegionType
            self.polygon: EVDemandInput.Polygon
            self.numCustomers: int
            if dict:
                self.__dict__ = dict
                self.shapely_polygon=shapely.Polygon(self.polygon.points)
    
    class VehicleUsage(Enum):
        Low=0
        Medium=1
        High=2

    class PredictorParams:
        def __init__(self, dict: None)->None:
            self.vehicleUsage: EVDemandInput.VehicleUsage
            self.years: list[int]
            if dict:
                self.__dict__ = dict
    
    @staticmethod
    def fromJson(jsonStr: str) -> 'EVDemandInput':
        return json.loads(jsonStr, object_hook=EVDemandInput._objectHook)
        
    @staticmethod
    def _objectHook(dict):
        if ( 'className' in dict):
            className = dict['className']
            if (className=='EVDemandInput'):
                return EVDemandInput(dict)
            elif (className=='EVDemandInput.RegionData'):            
                return EVDemandInput.RegionData(dict)
            elif (className=='EVDemandInput.PredictorParams'):
                return EVDemandInput.PredictorParams(dict)
            elif (className=='EVDemandInput.Polygon'):
                return EVDemandInput.Polygon(dict)

class EVDemands:
    def __init__(self) -> None:
        self.evDemandList:list[EVDemand]

class Quarter(Enum):
    Q1=0
    Q2=1
    Q3=2
    Q4=3

class EVDemand:
    def __init__(self) -> None:
        self.bev:DataValue
        self.phev:DataValue
        self.year:int
        self.quarter:Quarter

class DataValue:
    def __init__(self) -> None:
        self.mean:float = 0
        self.std:float = 0

File: CommonInterfaces/Models/EVDemandInput.cs
namespace CommonInterfaces.Models;

public class EVDemandInput {
    private void initialise() {
        predictorParams = new PredictorParams() { vehicleUsage = VehicleUsage.Medium};
        regionData = new List<RegionData>();
    }
    public EVDemandInput() {
        initialise();
    }

    public class Polygon {
        public Polygon() {

        }
        public string className {
            get {
                return "EVDemandInput.Polygon";
            }
        }
        public Polygon(double[] longitudes, double[] latitudes) {
            var minLength = Math.Min(latitudes.Length,longitudes.Length);
            points = new double[minLength][];
            for ( int i=0;i<minLength; i++) {
                points[i]=new double[2] { longitudes[i],latitudes[i]};
            }
        }
        public double[][] points {get; set;}
    }

    /// <summary>
    /// Defines the region over which the prediction will take place
    /// </summary> <summary>
    /// 
    /// </summary>
    public enum RegionType { Dist, Primary, GSP}
    public class RegionData {
        public RegionData() {

        }
        public RegionData(int _id, RegionType _type) {
            id = _id;
            type=_type;
        }
        public string className {
            get {
                return "EVDemandInput.RegionData";
            }
        }
        public int id {get; set;}
        public RegionType type{ get; set;}
        public Polygon polygon {get; set;}
        public int numCustomers {get; set;}
    } 
    /// <summary>
    /// Params associated with prediction
    /// </summary>
    public enum VehicleUsage { Low, Medium, High}
    public class PredictorParams {     
        public PredictorParams() {
            years = new List<int>();
        }       
        public string className {
            get {
                return "EVDemandInput.PredictorParams";
            }
        }
        public VehicleUsage vehicleUsage {get; set;}

        //?? Not used at present - but could be??
        public List<int> years {get; set;}
    } 

    public string className {
        get {
            return "EVDemandInput";
        }
    }
    public List<RegionData> regionData {get; set;}
    public PredictorParams predictorParams {get;set;}
}

